Match Screen Feature

M-Abdelnabi opened the pull request:
### **This PR introduces the Match Screen feature, which helps users discover movies tailored to their preferences.**

**Users can set filters based on:**

- Mood (e.g., excited, sad, emotional, etc.)

- Genres (e.g., action, comedy, drama, etc.)

- Duration (short, medium, long)

- Release Type: choose between Recent, Classic, or Both

**Based on these inputs, the system suggests a curated list of movies that best match the user’s choices.**


https://github.com/user-attachments/assets/d9bf7290-12bd-4148-8e61-380fc700cbe5

M-Abdelnabi committed:
feat: Add FlexboxLayout dependency

This commit introduces the `com.google.android.flexbox:flexbox:3.0.0` dependency to the `app/build.gradle` file. This library will be used for implementing match questions with flexible layouts.

M-Abdelnabi committed:
Fix: Corrected typos and updated drawables

- Corrected typos in content description and text in `item_tvshow_rating.xml` and `item_rating.xml`.
- Updated `rounded_corners.xml` to use `brand_color_tertiary` and set corner radius.
- Reordered attributes in `ic_magic_stick_selector.xml`.
- Updated path data for `ic_search.xml` to a new search icon.

M-Abdelnabi committed:
feat: Add new drawables for UI elements

This commit introduces a collection of new drawable resources.
These include:
- `circle_shape.xml`: A simple circular shape.
- `ic_flame.xml`: An icon representing a flame.
- `ic_headphones.xml`: An icon representing headphones.
- `ic_heart.xml`: An icon representing a heart.
- `ic_magic_wand.xml`: An icon representing a magic wand.
- `ic_time_long.xml`: An icon representing a longer duration.
- `ic_time_medium.xml`: An icon representing a medium duration.
- `ic_time_short.xml`: An icon representing a shorter duration.

M-Abdelnabi committed:
feat: add progress_drawable.xml

This commit introduces a new drawable resource for a custom progress bar.

It defines a layer-list with a background and progress item, both using rectangular shapes with rounded corners. The background is set to `card_background_color` and the progress indicator uses `brand_color_primary`.

M-Abdelnabi committed:
feat: Add drawables for question card background and rounded button

Added two new drawable resources:
- `question_card_background.xml`: A selector for question card backgrounds with different states (selected, pressed, default). It uses a rectangular shape with rounded corners and defines solid colors and strokes for each state.
- `rounded_button.xml`: A shape drawable for rounded buttons, using a rectangular shape with large radius corners and a primary button color.

M-Abdelnabi committed:
feat: add new strings and dimensions for Match Feature

This commit introduces new string resources for the "Match Feature", which helps users discover movies and series based on their preferences. It also includes new dimension values to support the UI of this feature. Additionally, it corrects typos in existing string resources.

M-Abdelnabi committed:
feat: Add match screen layout

This commit introduces the layout for the match feature, including:
- `fragment_match.xml`: The main layout file using data binding with `MatchViewModel`.
- It includes layouts for different states of the match process: `match_start_page`, `match_questions_page`, and `match_results_page`.
- An error layout `error2` is also included for displaying error states.

M-Abdelnabi committed:
feat: Add Match feature layouts

This commit introduces the layout files for the new "Match" feature.
- `match_start_page.xml`: Layout for the initial screen of the Match feature, prompting users to start the matching process.
- `match_questions_page.xml`: Layout for the questions screen where users provide their preferences (mood, genre, time, type). Includes a progress bar and sections for each question type.
- `match_results_page.xml`: Layout to display the movie recommendations. It features a carousel for movie posters and a details card for the selected movie.

M-Abdelnabi committed:
feat: Add new layouts for match feature

Added the following layout files:
- item_match_carousel.xml
- item_match_question.xml
- item_match_question_card.xml
- match_app_bar.xml

M-Abdelnabi committed:
Refactor: rename MyListFragment to MatchFragment

This commit renames `MyListFragment` to `MatchFragment` and updates its usage in navigation graphs and menu resources.

M-Abdelnabi committed:
feat: Add hide/show bottom navigation functionality

This commit introduces two new functions to `MainActivity.kt`:

- `hideBottomNavigation()`: Hides the bottom navigation view.
- `showBottomNavigation()`: Shows the bottom navigation view.

Additionally, `R.id.myListFragment` has been replaced with `R.id.matchFragment` in the `AppBarConfiguration` to reflect a change in the navigation graph.

M-Abdelnabi committed:
feat: add getMatchedMovies to MovieService

This commit introduces a new endpoint `getMatchedMovies` to the `MovieService` interface. This endpoint allows fetching a list of movies based on various filter criteria such as genres, runtime, and release date.

M-Abdelnabi committed:
feat: Add `getMatchedMovies` to `MovieRepository`

This commit introduces a new function `getMatchedMovies` to the `MovieRepository` interface and its implementation `MovieRepositoryImp`.

This function allows fetching a list of movies based on various matching criteria such as genres, runtime, and release date.

M-Abdelnabi committed:
feat: Created GetMatchedMoviesUseCase

This commit introduces a new use case, `GetMatchedMoviesUseCase`, responsible for fetching a list of movies that match specified criteria such as genre, runtime, and release date. It utilizes `MovieRepository` to retrieve data and `MovieMapper` to transform it into the domain model `Media`. Error handling is included to return an empty list in case of exceptions.

M-Abdelnabi committed:
feat: Add MatchMapper and MatchModels for movie matching feature

This commit introduces `MatchMapper.kt` and `MatchModels.kt`.

`MatchMapper.kt` provides functionality to convert UI state for movie matching (mood, genre, time, type) into parameters for fetching movie recommendations.

`MatchModels.kt` defines the data structures and enums used in the movie matching UI, including:
- `MatchPages` enum for different stages of the matching process.
- `QuestionType` enum for different categories of questions.
- `MatchUiState` data class to hold the overall UI state for matching.
- `QuestionUiState` and `GenreUiState` for individual question and genre options.
- `MatchParams` data class for parameters sent to retrieve recommendations.
- Helper functions to provide initial question sets for mood, movie type, time, and genre.

M-Abdelnabi committed:
feat: add match carousel and results adapters

M-Abdelnabi committed:
feat: add MatchOptionsAdapter

This commit introduces a new `MatchOptionsAdapter` for the movie match feature.

The adapter handles the display of different question options (e.g., mood, genre) in a RecyclerView. It supports dynamic question types and updates the UI accordingly. Each option can have a name, description, and icon. User selections are handled via a callback.

M-Abdelnabi committed:
feat: Add MatchViewModel and initial matching logic

This commit introduces the `MatchViewModel` responsible for handling the movie matching feature.

Key changes include:
- Added `MatchViewModel` with dependencies for getting matched movies and genre lists.
- Implemented logic to load movie genres on initialization.
- Added functions to handle UI events such as starting the matching process, navigating between questions, and selecting answers.
- Implemented logic to load movie recommendations based on user answers.
- Added error handling for failed API calls during recommendation loading.
- Introduced `MatchEvent` sealed class for UI events like movie clicks.
- Added navigation logic for back actions within the matching flow.
- Added a retry mechanism for loading matches.

M-Abdelnabi committed:
feat: Implement MatchFragment for movie matching feature

This commit introduces the `MatchFragment`, a new component responsible for guiding users through a series of questions to find movie recommendations.

Key functionalities implemented:
- **Multi-page UI:** Manages different states (Start, Questions, Results) for the matching process.
- **Questionnaire:** Dynamically displays questions related to mood, genre, time, and movie type.
- **Carousel Results:** Presents matched movies in an interactive carousel with auto-scroll and detailed view.
- **Navigation:** Handles back navigation and navigation to movie details.
- **State Management:** Utilizes a ViewModel to manage UI state and events.
- **Dynamic UI Updates:** Adapts the UI based on user selections and progress.
- **Error Handling:** Includes basic error display.

M-Abdelnabi committed:
Refactor: Removed unused navigation actions

The commit removes the navigation actions for `CreateButtonClicked` and `OnSelectItem` from `MyListsFragment`.

M-Abdelnabi committed:
Merge branch 'develop' into feature/match-screen

# Conflicts:
#	app/src/main/res/navigation/movie_navigation.xml
#	app/src/main/res/values/strings.xml

mohamedebrahem13 commented on app/src/main/res/values/strings.xml:
```diff
@@ -136,13 +136,59 @@
+    <string name="discover_your_match">Discover Your Match</string>
+    <string name="not_sure_watch">Not sure what to watch?</string>
+    <string name="answer_questions">Answer a few quick questions and we’ll\n match you with the perfect picks.</string>
+    <string name="start_matching">Start Matching</string>
```
change this text to be the first letter only capital not all the text "Start Matching"

M-Abdelnabi replied:
i'm just forget             
android:textAllCaps="false"
i resolve it

M-Abdelnabi committed:
feat: Add navigation action from MyListFragment to SavedListFragment

This commit introduces a new navigation action in `movie_navigation.xml`.

The action `action_myListFragment_to_savedListFragment` allows navigation from `MyListFragment` to `ListDetailsFragment`.

mohamedebrahem13 commented on app/src/main/java/com/karrar/movieapp/ui/match/MatchModels.kt:
```diff
@@ -0,0 +1,147 @@
+fun getMoodQuestionAnswers() = listOf(
+    QuestionUiState(
+        id = 1,
+        name = R.string.mood_chill,
```
the icon doesn't have background  i mean the icon after click on them the background disappear

M-Abdelnabi replied:
selector problem
solved 👍

M-Abdelnabi committed:
Refactor: Removed unused navigation action

The click listener for `buttonEmpty` in `MyListsFragment` was navigating to `ExploringFragment`, but this navigation was removed.

mohamedebrahem13 reviewed (changes requested):
grate work can you check please the icon why some of them doesn't have the background,
and check the string res for capital string should be small

M-Abdelnabi committed:
Merge branch 'develop' into feature/match-screen

mohamedebrahem13 commented on app/src/main/java/com/karrar/movieapp/ui/match/MatchModels.kt:
```diff
@@ -0,0 +1,147 @@
+    QuestionUiState(
+        id = 1,
+        name = R.string.time_short_label,
+        description = R.string.time_short_desc,
```
time icon not the same as design

M-Abdelnabi replied:
yes, i replace it

M-Abdelnabi committed:
Refactor: Removed unused string resource

The string "series" was removed from `strings.xml` as it was no longer in use.

M-Abdelnabi committed:
fix: update mood and time icons

- Update mood icons to new designs.
- Update time icons to new designs.
- Update time labels in strings.xml.

M-Abdelnabi committed:
Fix: Set textAllCaps to false for buttons

Set the `android:textAllCaps` attribute to `false` for buttons in the following layouts:
- `match_questions_page.xml`
- `match_start_page.xml`
- `match_results_page.xml`

This ensures that the button text is not automatically capitalized.

M-Abdelnabi committed:
Refactor: Updated rounded_corners.xml to use a selector for different states

The drawable now uses a selector to define different appearances for selected, pressed, and default states.

- Selected and pressed states use `brand_color_secondary`.
- The default state uses `brand_color_tertiary`.
- All states maintain a 10dp corner radius.
- A transparent stroke has been added to selected and default states.

mohamedebrahem13 reviewed (approved):
Great job !

mohamedebrahem13 commented:
I have small thing with the match result pages should be overlapped from right and left

M-Abdelnabi commented:
I tried put i couldn't get it
If u have an idea how could i apply it 
I'll make anew branch to fix it

mohamedebrahem13 commented:
i used this for the sider in home screen you need to calculate the screen size  

    private fun attachCarouselTransformer(viewPager: ViewPager2) {
        val displayMetrics = viewPager.resources.displayMetrics
        val screenWidth = displayMetrics.widthPixels

        // Minimum overlap in pixels (for small devices)
        val minSidePeek = viewPager.context.resources.getDimensionPixelOffset(R.dimen.spacing_extra_large)
        // Adaptive overlap: 5% of screen width or at least minSidePeek
        val sidePeek = maxOf(screenWidth * 0.05f, minSidePeek.toFloat())

        val extraLift = viewPager.context.resources.getDimensionPixelOffset(R.dimen.spacing_extra_extra_large)

        viewPager.setPageTransformer { page, position ->
            val offset = position * -sidePeek
            page.translationX = if (viewPager.layoutDirection == View.LAYOUT_DIRECTION_RTL) -offset else offset
            page.scaleX = 1f
            page.scaleY = 1f
            page.alpha = 0.8f + (1 - abs(position)) * 0.2f
            page.translationY = if (position == 0f) -extraLift.toFloat() else 0f
            page.translationZ = if (position == 0f) 1f else 0f
        }
    }



you can find it here at this link 
https://github.com/Shanghai-Squad/MovieApp/blob/feature/%2330-home-screen/app/src/main/java/com/karrar/movieapp/ui/home/adapter/HomeAdapter.kt

M-Abdelnabi commented:
I will see it , thanks mohamed❤️
